src/Workspaces/Remote/ServiceHub/Shared/ServerDirectStream.cs

```diff
@@ -0,0 +1,101 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.IO.Pipes;
+using System.Runtime.Remoting;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.CodeAnalysis.Remote
+{
+    /// <summary>
+    /// Direct stream between service hub server and client to pass around big chunk of data
+    /// </summary>
+    internal class ServerDirectStream : Stream
```

Comment:
It's still hidden.  You're still just giving the user a Stream (just like today).  

Right now, this PR and this feature is massively complex.  It's 6000+ lines added and introduces tons of complexity and concepts.  Please make it simpler and avoid complexity when not necessary.  We have very little runway to get this in, and more complexit just makes reviewing that much harder for functionality we are only speculating we need.